AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager

bdhirsh opened the pull request:
Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169295
* #169023
* __->__ #171353



cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

bdhirsh committed:
AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented:
Do you want review before the tests are fixed?

ezyang commented:
The PR desc looks reasonabel

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

bdhirsh commented:
tests should be fixed (the main change was I was tweaking the calling convention for activations and wasn't activations that needed to be `mark_dynamic`'d properly).

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -443,6 +445,24 @@ def get_proxy_slot(
     return res
 
 
+def _get_proxies(t: torch.Tensor) -> list[Proxy]:
```
I guess the important thing is you traverse into nested structure but return the proxies in no particular order. Kinda weird, but OK.

ezyang commented on torch/_functorch/_aot_autograd/descriptors.py:
```diff
@@ -747,3 +747,11 @@ class SavedForBackwardsAOTOutput(AOTOutput):
+
+
+@dataclasses.dataclass(frozen=True)
+class SavedForBackwardsNoVcCheckAOTOutput(AOTOutput):
```
Doc me

bdhirsh replied:
Done, also added a `Note [...]` style comment block that I ended up referencing in other bits of the PR

ezyang commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -4316,6 +4316,24 @@ def bwd(ctx, grad, x):
+            # for all tensors passed from fwd to bwd. Find which indices
+            # correspond to save_for_backward tensors.
+            for i, fwd_proxy in enumerate(bwd_freevars.keys()):
+                if fwd_proxy in saved_tensor_proxies:
```
verified deterministic use of set (but you might consider using OrderedSet anyway just for safety)

ezyang commented on torch/_functorch/partitioners.py:
```diff
@@ -979,13 +983,22 @@ def _extract_fwd_bwd_modules(
+    for i, node in enumerate(saved_values):
+        if node.meta.get("saved_tensor_with_no_vc_check", False):
+            no_vc_check_start_idx = i
+            break
```
Do you mind asserting that the suffix all have this meta set? Thanks!

ezyang commented on torch/_functorch/partitioners.py:
```diff
@@ -3025,6 +3052,20 @@ def min_cut_rematerialization_partition(
+            saved_values_no_vc_check.append(node)
+        else:
+            saved_values_with_vc_check.append(node)
+    saved_values = saved_values_with_vc_check + saved_values_no_vc_check
```
This seems annoyingly duped with the code above, what's going on?

bdhirsh replied:
You're right - the duping here is because each of the two main partitioner fn's has its own way of choosing saved_values (which I wanted to sort), but the duplication is ugly. I ended up moving the code into `_extract_fwd_bwd_modules` (which accepts the joint graph + list of `saved_values` as inputs), so it now accepts unordered saved_values and sorts them on the fly.

ezyang commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -2340,8 +2362,15 @@ def mark_dynamic_activations(activations: list[torch.Tensor]):
+                # 1. ctx.saved_tensors - tensors that went through save_for_backward (with VC check)
+                # 2. ctx._tensors_no_vc_check - tensors stashed directly on ctx (no VC check)
+                # The order matches the partitioner's output: VC-check tensors first, then no-VC-check tensors
+                all_saved_tensors = list(ctx.saved_tensors) + getattr(
```
How come getattr?

ezyang reviewed (approved):
All of the new bookkeeping is tiresome, but I think I agree there isn't a good way of fixing it without going straight to some sort of descriptor based approach.

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anobra commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -670,6 +677,42 @@ def tensors_saved_for_backwards_slice(self):
             return slice(self.num_forward, None)
 
+    @property
+    def tensors_saved_for_backwards_with_vc_check_slice(self):
```
now that you have separate slices for vc and no vc check, do you still need tensors_saved_for_backwards_slice? if so, would it make sense to implement tensors_saved_for_backwards_slice as concat of vc + no vc slices to make logic there explicit?

anobra commented on torch/_functorch/_aot_autograd/descriptors.py:
```diff
@@ -747,3 +747,30 @@ class SavedForBackwardsAOTOutput(AOTOutput):
+
+
+# Note [Activations with no version counter checks in eager]
+# In eager, when a tensor is saved or backward, the autograd engine
```
typo: *for backward

anobra reviewed (commented):
Nice! thanks for the fix.

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: at runtime, specialcase saved-for-bw tensors whos version counters werent checked in eager"

Partially authored with claude. See the test in `test/_dynamo/test_repros.py` for a good example.

Today if you have an autograd.Function, and you want to stash a tensor in the forward to be used in the backward, there are technically 2 ways to do so:

(1) `ctx.save_for_backward(x)`: this is the blessed way
(2) `ctx.x = x` This basically smuggles x into the backward, without the autograd engine knowing much about `x`. In particular, if you want to save `x` for backward, then if you mutate `x` later before running the backward, the autograd engine will run version counter checks if you used the `save_for_backward` API, but not if you do this smuggling.

This causes a parity issue between eager and compile: under compile, we always save stashed activations using `ctx.save_for_backward`. So if you happen to be relying on this eager behavior (say you stash an activation, mutate it *on purpose*, and then run the backward), then we will not error in eager as long as you stashed the tensor directly on `ctx`, but under compile we will error because we will generate code to save the tensor at runtime using `ctx.save_for_backward`.

This PR consists of a few pieces:

(1) In dynamo, tweak it's handling of `autograd.Function` so that it knows which saved tensors were saved using `ctx.save_for_backward` or not. Any tensors used in the backward that were **not** saved in this way get some extra node metadata attached to them

(2) When the partitioner decides on a set of activations to save, it figures out which activations have this metadata. For convenience (to make the list slicing easier), I tweaked the partitioner to group all activations into `activations_that_get_vc_checks, activations_that_dont_get_vc_checks`

(3) in AOTAutograd at runtime, we slice out these two sublists from the forward graph outputs separately. This allows us to only call `ctx.save_for_backward` on the "normal" activations, and directly stash the activations that shouldn't get VC checks on the ctx.




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
